### PR TITLE
Refactor image & label & prediction properties

### PR DESCRIPTION
**Pull request by matanper**
1. Refactored image properties to be the same structure as label & predictions properties
2. Refactored label & prediction measures to be called properties. 
3. Update label & prediction properties to accept batch and return a List (and not list of lists)
4. Change label & prediction property dict to have "output_type" which is continuous/discrete/class_id instead of "is_continuous"

### CODE REVIEW

File: deepchecks/vision/checks/distribution/image_dataset_drift.py [hunk 1]
```diff
@@ -38,9 +38,10 @@ class ImageDatasetDrift(TrainTestCheck):
 
     Parameters
     ----------
-    alternative_image_properties : List[str] , default: None
-        List of alternative image properties names. Must be attributes of the ImageFormatter classes that are passed to
-        train and test's VisionData class. If None, check uses DEFAULT_IMAGE_PROPERTIES.
+    alternative_image_properties : Dict[str, Callable] , default: None
+        Dict of image properties. Replaces the default deepchecks properties.
+        Each property have a name (key of dict) and function which accepts a `List[np.ndarray]` as input. If None, check
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/image_dataset_drift.py:43]:
```suggestion
        Each property has a name (key of dict) and function which accepts a `List[np.ndarray]` as input. If None, check
```

Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/image_dataset_drift.py:43]: Also, phrasing unclear. Would say that "each dictionary is comprised of name (key) and function (value)" or something of that sort

File: deepchecks/vision/checks/distribution/image_dataset_drift.py [hunk 2]
```diff
@@ -60,28 +61,28 @@ class ImageDatasetDrift(TrainTestCheck):
...
         self.n_top_properties = n_top_properties
         self.min_feature_importance = min_feature_importance
         self.sample_size = sample_size
         self.test_size = test_size
         self.min_meaningful_drift_score = min_meaningful_drift_score
 
-        self._train_properties = OrderedDict([(k, []) for k in self.image_properties])
-        self._test_properties = OrderedDict([(k, []) for k in self.image_properties])
+        self._train_properties = defaultdict(list)
+        self._test_properties = defaultdict(list)
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/image_dataset_drift.py:85]: Didn't know this, cool

File: deepchecks/vision/checks/distribution/image_property_drift.py [hunk 3]
```diff
@@ -43,35 +38,26 @@ class ImagePropertyDrift(TrainTestCheck):
...
     See https://en.wikipedia.org/wiki/Wasserstein_metric
 
-    Paramaters
+    Parameters
     ----------
-    image_properties : Optional[List[Union[str, Callable[..., Number]]]]
+    alternative_image_properties : Dict[str, Any], default: None
+        Dict of image properties. Replaces the default deepchecks properties.
+        Each property have a name (key of dict) and function which accepts a `List[np.ndarray]` as input. If None, check
+        uses `deepchecks.vision.utils.image_properties.default_image_properties`
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/image_property_drift.py:46]: Same comments

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 4]
```diff
@@ -24,21 +24,21 @@
...
     """
     Calculate label drift between train dataset and test dataset, using statistical measures.
 
     Check calculates a drift score for the label in test dataset, by comparing its distribution to the train
-    dataset. As the label may be complex, we run different measurements on the label and check their distribution.
+    dataset. As the label may be complex, we calculate different properties of the label and check their distribution.
 
-    A measurement on a label is any function that returns a single value or n-dimensional array of values. each value
-    represents a measurement on the label, such as number of objects in image or tilt of each bounding box in image.
+    A property on a label is any function that gets labels and returns list of numbers. each
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:38]:
```suggestion
    A label property is any function that gets labels and returns list of numbers. each
```

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 5]
```diff
@@ -24,21 +24,21 @@
...
+    number represents a property on the label, such as number of objects in image or tilt of each bounding box in image.
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:39]:
```suggestion
    number represents a property on a label, such as number of objects in image or tilt of each bounding box in image.
```

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 6]
```diff
@@ -55,77 +55,77 @@ class TrainTestLabelDrift(TrainTestCheck):
 
     Parameters
     ----------
-    alternative_label_measurements : List[Dict[str, Any]], default: None
-        List of measurements. Replaces the default deepchecks measurements.
-        Each measurement is dictionary with keys 'name' (str), 'method' (Callable) and 'is_continuous' (bool),
+    alternative_label_properties : List[Dict[str, Any]], default: None
+        List of properties. Replaces the default deepchecks properties.
+        Each property is dictionary with keys 'name' (str), 'method' (Callable) and 'value_type' (str),
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:60]: We should explain value types and their meaning

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 7]
```diff
@@ -135,29 +135,36 @@ def compute(self, context: Context) -> CheckResult:
...
 
             value, method, display = calc_drift_and_plot(
-                train_column=pd.Series(self._train_label_properties[measure_name]),
-                test_column=pd.Series(self._test_label_properties[measure_name]),
-                plot_title=measure_name,
-                column_type='numerical' if label_measure['is_continuous'] else 'categorical',
+                train_column=pd.Series(self._train_label_properties[name]),
+                test_column=pd.Series(self._test_label_properties[name]),
+                plot_title=name,
+                column_type='categorical' if value_type == 'class_id' else value_type,
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:153]: Didn't you say there's "class_id", "continuous" and "discrete"?

Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:153]: BTW maybe it should be "output_type"

File: deepchecks/vision/checks/methodology/simple_feature_contribution.py [hunk 8]
```diff
@@ -63,9 +62,10 @@ class SimpleFeatureContribution(TrainTestCheck):
 
     Parameters
     ----------
-    alternative_image_properties : List[str] , default: None
-        List of alternative image properties names. Must be attributes of the ImageFormatter classes that are passed to
-        train and test's VisionData class. If None, check uses default_image_properties.
+    alternative_image_properties : Dict[str, Any], default: None
+        Dict of image properties. Replaces the default deepchecks properties.
+        Each property have a name (key of dict) and function which accepts a `List[np.ndarray]` as input. If None, check
+        uses `deepchecks.vision.utils.image_properties.default_image_properties`
```
Comment (nirhutnik) [inline deepchecks/vision/checks/methodology/simple_feature_contribution.py:68]: Same comments

File: deepchecks/vision/utils/label_prediction_properties.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+from deepchecks.core.errors import DeepchecksValueError
+
+
+# Labels
+
+def _get_bbox_area(labels: List[torch.Tensor]) -> List[int]:
+    """Return a list containing the area of bboxes in batch."""
+    areas = [(label.reshape((-1, 5))[:, 4] * label.reshape((-1, 5))[:, 3]).tolist()
+             for label in labels]
```
Comment (nirhutnik) [inline deepchecks/vision/utils/label_prediction_properties.py:25]: Is this still the optimal way of getting the bbox area? Didn't we discuss a bbox class or something?

File: deepchecks/vision/utils/label_prediction_properties.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+DEFAULT_CLASSIFICATION_LABEL_PROPERTIES = [
+    {'name': 'Samples per class', 'method': _get_samples_per_class_classification, 'value_type': 'class'}
+]
+
+DEFAULT_OBJECT_DETECTION_LABEL_PROPERTIES = [
+    {'name': 'Samples per class', 'method': _get_samples_per_class_object_detection, 'value_type': 'class'},
+    {'name': 'Bounding box area (in pixels)', 'method': _get_bbox_area, 'value_type': 'numerical'},
+    {'name': 'Number of bounding boxes per image', 'method': _count_num_bboxes, 'value_type': 'numerical'},
+]
```
Comment (nirhutnik) [inline deepchecks/vision/utils/label_prediction_properties.py:54]: I know they are different, and that it makes sense, but I REALLY hate that label properties and image properties have a different format.

**Review by nirhutnik — APPROVED** [review]

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 7 above)
Comment (matanper) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:153]: I said numerical/categorical. Doesn't care much, but this are the names from tabular

File: deepchecks/vision/utils/label_prediction_properties.py (hunk 10 above)
Comment (matanper) [reply deepchecks/vision/utils/label_prediction_properties.py:54]: The main reason is image property are always numeric. But if we'll say "we support advance user properties for example weather (sunny/cloudy/rain/snow) then we can allow also categorical features and then we will need also there the value type

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 6 above)
Comment (matanper) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py:60]: in the user guide

File: deepchecks/vision/utils/label_prediction_properties.py (hunk 9 above)
Comment (matanper) [reply deepchecks/vision/utils/label_prediction_properties.py:25]: surely I haven't discussed that, maybe @noamzbr

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 11]
```diff
@@ -208,3 +215,9 @@ def condition(result: Dict) -> ConditionResult:
         return self.add_condition(f'PSI <= {max_allowed_psi_score} and Earth Mover\'s Distance <= '
                                   f'{max_allowed_earth_movers_score} for label drift',
                                   condition)
+
+
+def get_column_type(output_type):
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:220]: Shouldn't this be in some central location?

File: deepchecks/vision/checks/distribution/train_test_prediction_drift.py [hunk 12]
```diff
@@ -213,3 +215,9 @@ def condition(result: Dict) -> ConditionResult:
         return self.add_condition(f'PSI <= {max_allowed_psi_score} and Earth Mover\'s Distance <= '
                                   f'{max_allowed_earth_movers_score} for prediction drift',
                                   condition)
+
+
+def get_column_type(output_type):
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/train_test_prediction_drift.py:220]: again here - should have one of these

Comment (matanper) [reply deepchecks/vision/checks/distribution/train_test_prediction_drift.py]: done

File: deepchecks/vision/checks/distribution/train_test_label_drift.py (hunk 11 above)
Comment (matanper) [reply deepchecks/vision/checks/distribution/train_test_label_drift.py]: done

**Pull request merged**